lib/Instrumentation.lua

```diff
@@ -0,0 +1,183 @@
+local GlobalConfig = require(script.Parent.GlobalConfig)
+
+local Instrumentation = {}
+
+local ComponentStats = {}
+-- Tracks HELLA stats, including:
+	-- Recorded stats:
+		-- Render count by component
+		-- Update request count by component
+		-- Actual update count by component
+		-- shouldUpdate returned true count by component
+		-- Time taken to run shouldUpdate
+		-- Time taken to render by component
+	-- Derived stats:
+		-- Average render time by component
+		-- Percent of total render time by component
+		-- Average shouldUpdate time by component
+		-- Percent of time shouldUpdate returns true
+		-- Percent of total shouldUpdate time by component
+
+local function valuesByStat(t, f)
+	local a = {}
+	for _, obj in pairs(t) do table.insert(a, obj) end
+	table.sort(a, f)
+	local i = 0      -- iterator variable
+	local iter = function ()   -- iterator function
+		i = i + 1
+		if a[i] == nil then
+			return nil
+		else
+			return a[i]
+		end
+	end
+	return iter
+end
+
+local function getStatEntry(handle)
+	local name
+	if handle and handle._element and handle._element.component then
+		name = tostring(handle._element.component)
+	else
+		warn("Component name not valid for " .. tostring(handle._key))
+		return nil
+	end
+	local entry = ComponentStats[name]
+	if not entry then
+		-- Use shortened name for convenience; 
+		local shortName = name:gsub("Connection","Conn")
```

Comment:
Let's not shorten names like this - finding poorly-performing components could be tricky if the name gets squashed. If we absolutely must shorten names, we could try trimming the string after it passes a specific number of characters (maybe trim from the end, so `ReallyLongName` gets trimmed to something like `...ongName`? Not sure about the exact semantics).